File: apps/backend-gateway/src/interfaces/api/v1/healthRouter.ts
/**
 * StadiumOS AI — Health & Readiness Probes Router.
 * 
 * Exposes /health (Liveness) and /ready (Readiness) probes checking database
 * and Redis connectivity states.
 */

import { Router, Response, NextFunction } from "express";
import { prisma } from "../../../infrastructure/database/prisma";
import { getRedisClient } from "../../../infrastructure/cache/redisClient";
import { logger } from "../../../infrastructure/logging/logger";

const router = Router();

// Liveness probe (is process alive)
router.get(
  "/health",
  (req: any, res: Response) => {
    res.status(200).json({ status: "healthy", timestamp: new Date().toISOString() });
  }
);

// Readiness probe (is database and cache connected)
router.get(
  "/ready",
  async (req: any, res: Response) => {
    const checks: Record<string, string> = {
      database: "down",
      cache: "down"
    };
    
    let isReady = true;

    // 1. Inspect Database
    try {
      await prisma.$queryRaw`SELECT 1`;
      checks.database = "up";
    } catch (err: any) {
      logger.error("Readiness check database failure:", err.message);
      isReady = false;
    }

    // 2. Inspect Redis Cache
    try {
      const redis = getRedisClient();
      const ping = await redis.ping();
      if (ping === "PONG") {
        checks.cache = "up";
      } else {
        isReady = false;
      }
    } catch (err: any) {
      logger.error("Readiness check cache failure:", err.message);
      isReady = false;
    }

    const statusCode = isReady ? 200 : 503;
    res.status(statusCode).json({
      status: isReady ? "ready" : "not_ready",
      components: checks,
      timestamp: new Date().toISOString()
    });
  }
);

export { router as healthRouter };
export default router;

File: apps/backend-gateway/src/interfaces/api/v1/health_router.py
"""
StadiumOS AI — Health & Readiness API Router.

Exposes endpoints for monitoring container status.
Used by container orchestrators (Kubernetes) to verify pod liveness and readiness.
"""

from fastapi import APIRouter, Depends, status, Response
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession
import structlog

from src.infrastructure.database.engine import get_db_session, get_engine
from src.infrastructure.cache.redis_client import get_redis_client

router = APIRouter(tags=["Health"])
logger = structlog.get_logger("health_router")


@router.get(
    "/health",
    status_code=status.HTTP_200_OK,
    summary="Liveness Probe",
)
async def liveness_probe() -> dict[str, str]:
    """
    Verify the application process is running and responsive.
    """
    return {"status": "healthy"}


@router.get(
    "/ready",
    summary="Readiness Probe",
)
async def readiness_probe(
    response: Response,
    db: AsyncSession = Depends(get_db_session),
) -> dict[str, Any] if False else dict:
    """
    Verify the application can serve requests by checking database and cache health.

    Returns HTTP 503 if any dependency is down, signaling Kubernetes to stop
    routing traffic to this container instance.
    """
    db_healthy = False
    redis_healthy = False
    
    # 1. Test Database Session
    try:
        await db.execute(text("SELECT 1"))
        db_healthy = True
    except Exception as e:
        await logger.aerror("readiness_db_check_failed", error=str(e))

    # 2. Test Redis connection
    try:
        redis_client = get_redis_client()
        await redis_client.ping()
        redis_healthy = True
    except Exception as e:
        await logger.aerror("readiness_redis_check_failed", error=str(e))

    if not db_healthy or not redis_healthy:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
        return {
            "status": "unavailable",
            "checks": {
                "database": "up" if db_healthy else "down",
                "redis": "up" if redis_healthy else "down",
            }
        }

    return {
        "status": "ready",
        "checks": {
            "database": "up",
            "redis": "up",
        }
    }
